src/event-stream/event-stream.service.ts

```diff
@@ -0,0 +1,191 @@
+    private handleReceipt: (receipt: EventStreamReply) => void,
+  ) {
+    this.init();
+  }
+
+  private init() {
+    this.disconnectDetected = false;
+    this.closeRequested = false;
+
+    this.ws = new WebSocket(this.url);
+    this.ws
+      .on('open', () => {
+        if (this.disconnectDetected) {
+          this.disconnectDetected = false;
+          this.logger.log('Event stream websocket restored');
+        } else {
+          this.logger.log('Event stream websocket connected');
+        }
+        this.produce({ type: 'listen', topic: this.topic });
+        this.produce({ type: 'listenreplies' });
+        this.ping();
+      })
+      .on('close', () => {
+        if (this.closeRequested) {
+          this.logger.log('Event stream websocket closed');
+        } else {
+          this.disconnectDetected = true;
+          this.logger.error(
+            `Event stream websocket disconnected, attempting to reconnect in ${RECONNECT_TIME}ms`,
+          );
+          setTimeout(() => this.init(), RECONNECT_TIME);
+        }
+      })
+      .on('message', (message: string) => {
+        this.handleMessage(JSON.parse(message));
+      })
+      .on('pong', () => {
+        this.ping();
+      })
+      .on('error', err => {
+        this.logger.error(`Event stream websocket error: ${err}`);
+      });
+  }
+
+  private ping() {
+    this.ws.ping();
+    clearTimeout(this.pingTimeout);
+    this.pingTimeout = setTimeout(() => {
+      this.logger.error('Event stream ping timeout');
+      this.ws.terminate();
+    }, PING_TIMEOUT);
+  }
+
+  private produce(message: any) {
+    this.ws.send(JSON.stringify(message));
+  }
+
+  ack() {
+    this.produce({ type: 'ack', topic: this.topic });
+  }
+
+  close() {
+    this.closeRequested = true;
+    this.ws.terminate();
+  }
+
+  private handleMessage(message: EventStreamReply | Event[]) {
+    if (Array.isArray(message)) {
+      for (const event of message) {
+        this.logger.log(`Ethconnect '${event.signature}' message: ${JSON.stringify(event.data)}`);
+      }
+      this.handleEvents(message);
+    } else {
+      const replyType = message.headers.type;
+      const errorMessage = message.errorMessage ?? '';
+      this.logger.log(
+        `Ethconnect '${replyType}' reply request=${message.headers.requestId} tx=${message.transactionHash} ${errorMessage}`,
+      );
+      this.handleReceipt(message);
+    }
+  }
+}
+@Injectable()
+export class EventStreamService {
+  private readonly logger = new Logger(EventStreamService.name);
+
+  constructor(private http: HttpService) {}
+
+  async ensureEventStream(baseUrl: string, topic: string): Promise<EventStream> {
+    const streamDetails = {
+      name: topic,
+      errorHandling: 'block',
+      batchSize: 5,
+      batchTimeoutMS: 500,
+      type: 'websocket',
+      websocket: { topic },
+      blockedReryDelaySec: 30, // intentional due to spelling error in ethconnect
+    };
```

Comment:
Note that these stream details are completely non-configurable from the client (ie FireFly). Hopefully that's alright, at least for an initial pass.

Reply:
Yes, I think that's completely correct.
If we introduced tuning options, I would expect it to go into the config of this service (not into FireFly).
That's certainly the precedent set by other microservice style connectors/plugins, like DX. The FireFly config just tells FireFly how to connect to the remote connector. The config within that connector has all the detailed stuff specific to its operation.

I note it would be totally fine for someone to create a peer tokens microservice that doesn't use Ethconnect at all, and manages all the event listening internally with direct `web3` calls. Potentially as long as the REST API from FireFly<->Tokens stays the same, the developer of that Node.js microservice wouldn't need to make any changes at all to the FireFly Go code to add another plugin that side, or extra config options. So I think the separation of concerns that pushes specific configuration details down into the Microservice is good for future extensibility.